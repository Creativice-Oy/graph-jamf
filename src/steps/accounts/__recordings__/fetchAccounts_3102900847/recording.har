{
  "log": {
    "_recordingName": "fetchAccounts",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "4f0e505db41b492155ff6d75affa0642",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "identity"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiteronedev.jamfcloud.com"
            }
          ],
          "headersSize": 364,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "queryString": [],
          "url": "https://jupiteronedev.jamfcloud.com/api/v1/auth/token"
        },
        "response": {
          "bodySize": 392,
          "content": {
            "mimeType": "application/json",
            "size": 392,
            "text": "{\"token\":\"[REDACTED]\",\"expires\":\"2023-01-18T20:30:08.913Z\"}"
          },
          "cookies": [
            {
              "expires": "2023-01-25T20:00:08.000Z",
              "name": "AWSALB",
              "path": "/",
              "value": "[REDACTED]"
            },
            {
              "expires": "2023-01-25T20:00:08.000Z",
              "name": "AWSALBCORS",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "expires": "1970-01-01T00:00:00.000Z",
              "httpOnly": true,
              "maxAge": 0,
              "name": "JSESSIONID",
              "path": "/",
              "sameSite": "Lax",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "APBALANCEID",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 18 Jan 2023 20:00:08 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "Jamf Cloud Node"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains;, max-age=31536000 ; includeSubDomains"
            },
            {
              "name": "expires",
              "value": "0"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-frame-options",
              "value": "DENY"
            }
          ],
          "headersSize": 969,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-01-18T20:00:08.345Z",
        "time": 670,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 670
        }
      },
      {
        "_id": "e0cc2e06750b002c84a3dde729e7f689",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "identity"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiteronedev.jamfcloud.com"
            }
          ],
          "headersSize": 639,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://jupiteronedev.jamfcloud.com/api/v1/jamf-pro-version"
        },
        "response": {
          "bodySize": 39,
          "content": {
            "mimeType": "application/json",
            "size": 39,
            "text": "{\"version\":\"10.42.1-t1667311080\"}"
          },
          "cookies": [
            {
              "expires": "2023-01-25T20:00:09.000Z",
              "name": "AWSALB",
              "path": "/",
              "value": "[REDACTED]"
            },
            {
              "expires": "2023-01-25T20:00:09.000Z",
              "name": "AWSALBCORS",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "APBALANCEID",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 18 Jan 2023 20:00:09 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "Jamf Cloud Node"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains;, max-age=31536000 ; includeSubDomains"
            },
            {
              "name": "expires",
              "value": "0"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store, must-revalidate"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-frame-options",
              "value": "DENY"
            }
          ],
          "headersSize": 855,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-01-18T20:00:09.024Z",
        "time": 434,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 434
        }
      },
      {
        "_id": "25958b0444ec2a193681bde268b4670d",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "identity"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiteronedev.jamfcloud.com"
            }
          ],
          "headersSize": 636,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://jupiteronedev.jamfcloud.com/JSSResource/accounts"
        },
        "response": {
          "bodySize": 692,
          "content": {
            "mimeType": "text/plain;charset=UTF-8",
            "size": 692,
            "text": "{\"accounts\":{\"users\":[{\"id\":12,\"name\":\"adam.pierson\"},{\"id\":1,\"name\":\"adam.williams\"},{\"id\":9,\"name\":\"austin.kelleher\"},{\"id\":13,\"name\":\"chasen.bettinger\"},{\"id\":3,\"name\":\"erkang.zheng\"},{\"id\":7,\"name\":\"isaacwilliams\"},{\"id\":8,\"name\":\"jupiterone.dev\"},{\"id\":14,\"name\":\"max.daum\"},{\"id\":17,\"name\":\"my.new.account\"},{\"id\":10,\"name\":\"nick.dowmon\"},{\"id\":15,\"name\":\"samuel.poulton\"},{\"id\":11,\"name\":\"stefan.virag\"},{\"id\":16,\"name\":\"Wrong_Permissions\"}],\"groups\":[{\"id\":1,\"name\":\"Test Group\",\"site\":{\"id\":1,\"name\":\"Test site\"}},{\"id\":2,\"name\":\"Another Test Group\",\"site\":{\"id\":1,\"name\":\"Test site\"}},{\"id\":3,\"name\":\"Not all groups have members inside in api\",\"site\":{\"id\":1,\"name\":\"Test site\"}}]}}"
          },
          "cookies": [
            {
              "expires": "2023-01-25T20:00:09.000Z",
              "name": "AWSALB",
              "path": "/",
              "value": "[REDACTED]"
            },
            {
              "expires": "2023-01-25T20:00:09.000Z",
              "name": "AWSALBCORS",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "APBALANCEID",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 18 Jan 2023 20:00:09 GMT"
            },
            {
              "name": "content-type",
              "value": "text/plain;charset=UTF-8"
            },
            {
              "name": "content-length",
              "value": "692"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "Jamf Cloud Node"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubdomains;, max-age=31536000 ; includeSubDomains"
            },
            {
              "name": "x-frame-options",
              "value": "DENY"
            },
            {
              "name": "cache-control",
              "value": "no-store, no-cache, must-revalidate, max-age=0, post-check=0, pre-check=0"
            },
            {
              "name": "accept-ranges",
              "value": "bytes"
            },
            {
              "name": "vary",
              "value": "Accept-Charset,Accept-Encoding,Accept-Language,Accept"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 964,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2023-01-18T20:00:09.498Z",
        "time": 477,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 477
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
