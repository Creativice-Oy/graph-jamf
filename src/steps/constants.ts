import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const ACCOUNT_ENTITY_KEY = 'ACCOUNT_ENTITY';
export const ACCOUNT_DATA_KEY = 'ACCOUNT_DATA';
export const MAC_OS_CONFIGURATION_DETAILS_BY_ID_KEY =
  'MAC_OS_CONFIGURATION_DETAILS_BY_ID';

export enum IntegrationSteps {
  ACCOUNTS = 'fetch-accounts',
  MACOS_CONFIGURATION_PROFILES = 'fetch-macos-configuration-profiles',
  MOBILE_DEVICES = 'fetch-mobile-devices',
  COMPUTERS = 'fetch-computers',
  GROUPS = 'fetch-groups',
  ADMINS = 'fetch-admins',
  DEVICE_USERS = 'fetch-device-users'
}

export const Entities: Record<
  | 'ACCOUNT'
  | 'GROUP'
  | 'USER_ADMIN'
  | 'DEVICE_USER'
  | 'MOBILE_DEVICE'
  | 'COMPUTER'
  | 'MAC_OS_CONFIGURATION_PROFILE'
  | 'MAC_OS_APPLICATION',
  StepEntityMetadata
> = {
  ACCOUNT: {
    _type: 'jamf_account',
    _class: ['Account'],
    resourceName: 'Account',
  },
  GROUP: {
    _type: 'jamf_group',
    _class: ['UserGroup'],
    resourceName: 'Group',
  },
  USER_ADMIN: {
    _type: 'jamf_user',
    _class: ['User'],
    resourceName: 'Admin',
  },
  DEVICE_USER: {
    _type: 'device_user',
    _class: ['User'],
    resourceName: 'User',
  },
  MOBILE_DEVICE: {
    _type: 'mobile_device',
    _class: ['Device'],
    resourceName: 'Mobile Device',
  },
  COMPUTER: {
    _type: 'user_endpoint',
    _class: ['Host', 'Device'],
    resourceName: 'Computer',
  },
  MAC_OS_CONFIGURATION_PROFILE: {
    _type: 'jamf_osx_configuration_profile',
    _class: ['Configuration'],
    resourceName: 'macOS Configuration Profile',
  },
  MAC_OS_APPLICATION: {
    _type: 'macos_app',
    _class: ['Application'],
    resourceName: 'macOS Application',
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_USER_ADMIN'
  | 'ACCOUNT_HAS_GROUP'
  | 'ACCOUNT_HAS_DEVICE_USER'
  | 'ACCOUNT_HAS_MOBILE_DEVICE'
  | 'ACCOUNT_HAS_COMPUTER'
  | 'ACCOUNT_HAS_MAC_OS_CONFIGURATION_PROFILE'
  | 'COMPUTER_INSTALLED_APPLICATION'
  | 'COMPUTER_USES_PROFILE'
  | 'DEVICE_USER_HAS_MOBILE_DEVICE'
  | 'DEVICE_USER_HAS_COMPUTER'
  | 'GROUP_HAS_USER_ADMIN_MEMBER',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USER_ADMIN: {
    _type: 'jamf_account_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.USER_ADMIN._type,
  },
  ACCOUNT_HAS_GROUP: {
    _type: 'jamf_account_has_group',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.GROUP._type,
  },
  ACCOUNT_HAS_DEVICE_USER: {
    _type: 'jamf_account_has_device_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.DEVICE_USER._type,
  },
  ACCOUNT_HAS_MOBILE_DEVICE: {
    _type: 'jamf_account_has_mobile_device',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.MOBILE_DEVICE._type,
  },
  ACCOUNT_HAS_COMPUTER: {
    _type: 'jamf_account_has_user_endpoint',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.COMPUTER._type,
  },
  ACCOUNT_HAS_MAC_OS_CONFIGURATION_PROFILE: {
    _type: 'jamf_account_has_osx_configuration_profile',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.MAC_OS_CONFIGURATION_PROFILE._type,
  },
  COMPUTER_USES_PROFILE: {
    _type: 'user_endpoint_uses_osx_configuration_profile',
    _class: RelationshipClass.USES,
    sourceType: Entities.COMPUTER._type,
    targetType: Entities.MAC_OS_CONFIGURATION_PROFILE._type,
  },
  COMPUTER_INSTALLED_APPLICATION: {
    _type: 'user_endpoint_installed_application',
    _class: RelationshipClass.INSTALLED,
    sourceType: Entities.COMPUTER._type,
    targetType: Entities.MAC_OS_APPLICATION._type,
  },
  DEVICE_USER_HAS_MOBILE_DEVICE: {
    _type: 'device_user_has_mobile_device',
    _class: RelationshipClass.HAS,
    sourceType: Entities.DEVICE_USER._type,
    targetType: Entities.MOBILE_DEVICE._type,
  },
  DEVICE_USER_HAS_COMPUTER: {
    _type: 'device_user_has_user_endpoint',
    _class: RelationshipClass.HAS,
    sourceType: Entities.DEVICE_USER._type,
    targetType: Entities.COMPUTER._type,
  },
  GROUP_HAS_USER_ADMIN_MEMBER: {
    _type: 'jamf_group_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.GROUP._type,
    targetType: Entities.USER_ADMIN._type,
  },
};
