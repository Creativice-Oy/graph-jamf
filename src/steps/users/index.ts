import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  JobState,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { createClient, JamfClient } from '../../jamf/client';
import { IntegrationConfig } from '../../config';
import { Entities, Relationships } from '../constants';
import { createAdminEntity, createDeviceUserEntity } from './converters';
import { Admin, User } from '../../jamf/types';
import { getAccountData, getAccountEntity } from '../../util/account';
import { generateEntityKey } from '../../util/generateKey';

async function iterateAdminUserProfiles(
  client: JamfClient,
  jobState: JobState,
  iteratee: (user: Admin) => Promise<void>,
) {
  const accountData = await getAccountData(jobState);

  for (const user of accountData.users) {
    await iteratee(await client.fetchAccountUserById(user.id));
  }
}

async function iterateUserProfiles(
  client: JamfClient,
  iteratee: (user: User) => Promise<void>,
) {
  const users = await client.fetchUsers();

  for (const user of users) {
    await iteratee(await client.fetchUserById(user.id));
  }
}

async function createUserHasMobileDeviceRelationships(
  jobState: JobState,
  userEntity: Entity,
  user: User,
) {
  for (const mobileDevice of user.links?.mobile_devices || []) {
    const mobileDeviceEntity = await jobState.findEntity(
      generateEntityKey(Entities.MOBILE_DEVICE._type, mobileDevice.id),
    );

    if (!mobileDeviceEntity) {
      continue;
    }

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: userEntity,
        to: mobileDeviceEntity,
      }),
    );
  }
}

async function createUserHasComputerDeviceRelationships(
  jobState: JobState,
  userEntity: Entity,
  user: User,
) {
  for (const computerDevice of user.links?.computers || []) {
    const computerDeviceEntity = await jobState.findEntity(
      generateEntityKey(Entities.COMPUTER._type, computerDevice.id),
    );

    if (!computerDeviceEntity) {
      continue;
    }

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: userEntity,
        to: computerDeviceEntity,
      }),
    );
  }
}

export async function fetchAdminUsers({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const { config } = instance;

  const client = createClient({
    host: config.jamfHost,
    username: config.jamfUsername,
    password: config.jamfPassword,
    logger,
  });

  const accountEntity = await getAccountEntity(jobState);

  await iterateAdminUserProfiles(client, jobState, async (user) => {
    const adminEntity = await jobState.addEntity(createAdminEntity(user));

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: adminEntity,
      }),
    );
  });
}

export async function fetchDeviceUsers({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const { config } = instance;

  const client = createClient({
    host: config.jamfHost,
    username: config.jamfUsername,
    password: config.jamfPassword,
    logger,
  });

  const accountEntity = await getAccountEntity(jobState);

  await iterateUserProfiles(client, async (user) => {
    const userEntity = await jobState.addEntity(createDeviceUserEntity(user));

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );

    await createUserHasMobileDeviceRelationships(jobState, userEntity, user);

    await createUserHasComputerDeviceRelationships(jobState, userEntity, user);
  });
}

export const userSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-admins',
    name: 'Fetch Admins',
    entities: [Entities.USER_ADMIN],
    relationships: [Relationships.ACCOUNT_HAS_USER_ADMIN],
    executionHandler: fetchAdminUsers,
    dependsOn: ['fetch-accounts'],
  },
  {
    id: 'fetch-device-users',
    name: 'Fetch Device Users',
    entities: [Entities.DEVICE_USER],
    relationships: [
      Relationships.ACCOUNT_HAS_DEVICE_USER,
      Relationships.DEVICE_USER_HAS_MOBILE_DEVICE,
      Relationships.DEVICE_USER_HAS_COMPUTER,
    ],
    executionHandler: fetchDeviceUsers,
    dependsOn: ['fetch-accounts', 'fetch-mobile-devices', 'fetch-computers'],
  },
];
