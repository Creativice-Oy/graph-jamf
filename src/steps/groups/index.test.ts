import { fetchGroups } from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { fetchAccounts } from '../accounts';
import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { fetchAdminUsers } from '../users';

jest.setTimeout(5000000);

describe('#fetchGroups', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchGroups',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchAccounts, fetchAdminUsers, fetchGroups],
      entitySchemaMatchers: [
        {
          _type: Entities.GROUP._type,
          matcher: {
            _class: ['UserGroup'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'jamf_group' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                accessLevel: { type: 'string' },
                privilegeSet: { type: 'string' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.ACCOUNT_HAS_GROUP._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.HAS },
                _type: { const: 'jamf_account_has_group' },
              },
            },
          },
        },
        {
          _type: Relationships.GROUP_HAS_USER_ADMIN_MEMBER._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.HAS },
                _type: { const: 'jamf_group_has_user' },
              },
            },
          },
        },
      ],
    });
  }, 10000);
});
