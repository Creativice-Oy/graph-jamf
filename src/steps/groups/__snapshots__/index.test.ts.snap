// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchGroups should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "jamf_account_local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "local-integration-instance",
            "name": "Local Integration",
          },
        },
      ],
      "_type": "jamf_account",
      "createdOn": undefined,
      "displayName": "Local Integration",
      "id": "local-integration-instance",
      "name": "Local Integration",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "jamf_user_1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_user",
      "accessLevel": "Full Access",
      "active": true,
      "admin": true,
      "createdOn": undefined,
      "directoryUser": false,
      "displayName": "adam.williams",
      "email": "adam.williams@jupiterone.com",
      "emailAddress": "adam.williams@jupiterone.com",
      "enabled": "Enabled",
      "forcePasswordChange": false,
      "fullName": "",
      "id": "1",
      "name": "adam.williams",
      "permissions": Array [
        "Administrator",
      ],
      "privilegeSet": "Administrator",
      "username": "adam.williams",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "jamf_user_9",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_user",
      "accessLevel": "Full Access",
      "active": true,
      "admin": true,
      "createdOn": undefined,
      "directoryUser": false,
      "displayName": "austin.kelleher",
      "email": "austin.kelleher@jupiterone.com",
      "emailAddress": "austin.kelleher@jupiterone.com",
      "enabled": "Enabled",
      "forcePasswordChange": false,
      "fullName": "Austin Kelleher",
      "id": "9",
      "name": "austin.kelleher",
      "permissions": Array [
        "Administrator",
      ],
      "privilegeSet": "Administrator",
      "username": "austin.kelleher",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "jamf_user_3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_user",
      "accessLevel": "Full Access",
      "active": true,
      "admin": true,
      "createdOn": undefined,
      "directoryUser": false,
      "displayName": "erkang.zheng",
      "email": "erkang.zheng@lifeomic.com",
      "emailAddress": "erkang.zheng@lifeomic.com",
      "enabled": "Enabled",
      "forcePasswordChange": false,
      "fullName": "Erkang Zheng",
      "id": "3",
      "name": "erkang.zheng",
      "permissions": Array [
        "Administrator",
      ],
      "privilegeSet": "Administrator",
      "username": "erkang.zheng",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "jamf_user_7",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_user",
      "accessLevel": "Full Access",
      "active": true,
      "admin": true,
      "createdOn": undefined,
      "directoryUser": false,
      "displayName": "isaacwilliams",
      "email": "isaac.williams@lifeomic.com",
      "emailAddress": "isaac.williams@lifeomic.com",
      "enabled": "Enabled",
      "forcePasswordChange": false,
      "fullName": "Isaac Williams",
      "id": "7",
      "name": "isaacwilliams",
      "permissions": Array [
        "Administrator",
      ],
      "privilegeSet": "Administrator",
      "username": "isaacwilliams",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "jamf_user_8",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_user",
      "accessLevel": "Full Access",
      "active": true,
      "admin": true,
      "createdOn": undefined,
      "directoryUser": false,
      "displayName": "jupiterone.dev",
      "email": undefined,
      "emailAddress": "",
      "enabled": "Enabled",
      "forcePasswordChange": false,
      "fullName": "JupiterOne API Development",
      "id": "8",
      "name": "jupiterone.dev",
      "permissions": undefined,
      "privilegeSet": "Custom",
      "username": "jupiterone.dev",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "jamf_group_1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_group",
      "accessLevel": "Site Access",
      "createdOn": undefined,
      "displayName": "Test Group",
      "id": "1",
      "name": "Test Group",
      "privilegeSet": "Custom",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "jamf_group_2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_group",
      "accessLevel": "Site Access",
      "createdOn": undefined,
      "displayName": "Another Test Group",
      "id": "2",
      "name": "Another Test Group",
      "privilegeSet": "Custom",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "jamf_group_3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data currently disabled for this entity type",
          },
        },
      ],
      "_type": "jamf_group",
      "accessLevel": "Site Access",
      "createdOn": undefined,
      "displayName": "Not all groups have members inside in api",
      "id": "3",
      "name": "Not all groups have members inside in api",
      "privilegeSet": "Custom",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_user_1",
      "_toEntityKey": "jamf_user_1",
      "_type": "jamf_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_user_9",
      "_toEntityKey": "jamf_user_9",
      "_type": "jamf_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_user_3",
      "_toEntityKey": "jamf_user_3",
      "_type": "jamf_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_user_7",
      "_toEntityKey": "jamf_user_7",
      "_type": "jamf_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_user_8",
      "_toEntityKey": "jamf_user_8",
      "_type": "jamf_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_group_1",
      "_toEntityKey": "jamf_group_1",
      "_type": "jamf_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_group_2",
      "_toEntityKey": "jamf_group_2",
      "_type": "jamf_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "jamf_account_local-integration-instance",
      "_key": "jamf_account_local-integration-instance|has|jamf_group_3",
      "_toEntityKey": "jamf_group_3",
      "_type": "jamf_account_has_group",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "jamf_account",
    "jamf_user",
    "jamf_account_has_user",
    "jamf_group",
    "jamf_account_has_group",
  ],
  "numCollectedEntities": 9,
  "numCollectedRelationships": 8,
}
`;
